### PR TITLE
Add join function

### PR DESCRIPTION
**Pull request by j3r3m1**
Need to be approved before to modify the processing chain

### CODE REVIEW

File: geoindicators/src/main/groovy/org/orbisgis/orbisprocess/geoclimate/geoindicators/SpatialUnits.groovy [hunk 1]
```diff
@@ -351,53 +351,85 @@ IProcess createBlocks() {
...
+                datasource: JdbcDataSource
+        outputs outputTableName: String, idColumnTab2: String
+        run { inputTable1, inputTable2, idColumnTab1, idColumnTab2, prefixName, pointOnSurface,
+              nbRelations, datasource ->
 
             def GEOMETRIC_COLUMN_LOW = "the_geom"
             def GEOMETRIC_COLUMN_UP = "the_geom"
 
-            info "Creating the Tables of relations between the twos scales :  $inputLowerScaleTableName and $inputUpperScaleTableName"
+            info "Creating a spatial join between objects from two tables :  $inputTable1 and $inputTable2"
```
Comment (ebocher) [inline geoindicators/src/main/groovy/org/orbisgis/orbisprocess/geoclimate/geoindicators/SpatialUnits.groovy:390]: Creating a spatial join between geometries from two tables :  $inputTable1 and $inputTable2

File: geoindicators/src/main/groovy/org/orbisgis/orbisprocess/geoclimate/geoindicators/SpatialUnits.groovy [hunk 2]
```diff
@@ -351,53 +351,85 @@ IProcess createBlocks() {
...
+            def outputTableName = postfix "${inputTable1}_${inputTable2}", "join"
+            datasource."$inputTable1".the_geom.createIndex()
+            datasource."$inputTable2".the_geom.createIndex()
+            datasource."$inputTable2"[idColumnTab2].createIndex()
+
+            def query = """DROP TABLE IF EXISTS $outputTableName;"""
+            def subqueryNbRelations = ""
+            if (pointOnSurface){
+                query += """CREATE TABLE $outputTableName AS SELECT a.*, b.$idColumnTab2 
+                                    FROM $inputTable1 a, $inputTable2 b
```
Comment (ebocher) [inline geoindicators/src/main/groovy/org/orbisgis/orbisprocess/geoclimate/geoindicators/SpatialUnits.groovy:403]: CREATE TABLE $outputTableName AS SELECT a.*, b.$idColumnTab2 
                                    FROM $inputTable1 a, $inputTable2 b 
                                    WHERE   a.$GEOMETRIC_COLUMN_LOW && b.$GEOMETRIC_COLUMN_UP AND 
                                            ST_INTERSECTS(ST_POINTONSURFACE(a.$GEOMETRIC_COLUMN_LOW), st_force2d(b.$GEOMETRIC_COLUMN_UP))

File: geoindicators/src/main/groovy/org/orbisgis/orbisprocess/geoclimate/geoindicators/SpatialUnits.groovy [hunk 3]
```diff
@@ -351,53 +351,85 @@ IProcess createBlocks() {
...
+                query += """DROP TABLE IF EXISTS $temporaryTab;
+                            CREATE TABLE $temporaryTab 
+                                    AS SELECT   a.*, b.$idColumnTab2,
+                                                ST_AREA(ST_INTERSECTION(st_force2d(st_makevalid(a.$GEOMETRIC_COLUMN_LOW)), 
+                                                st_force2d(st_makevalid(b.$GEOMETRIC_COLUMN_UP)))) AS AREA
+                                    FROM    $inputTable1 a, $inputTable2 b
+                                    WHERE   a.$GEOMETRIC_COLUMN_LOW && b.$GEOMETRIC_COLUMN_UP AND 
+                                            ST_INTERSECTS(st_force2d(a.$GEOMETRIC_COLUMN_LOW), st_force2d(b.$GEOMETRIC_COLUMN_UP));
+                            CREATE INDEX IF NOT EXISTS id_area ON $temporaryTab USING BTREE(AREA);
+                            CREATE TABLE $outputTableName
```
Comment (ebocher) [inline geoindicators/src/main/groovy/org/orbisgis/orbisprocess/geoclimate/geoindicators/SpatialUnits.groovy:421]: Use the same approach as  :   

```
 datasource """DROP TABLE IF EXISTS $outputTableName;
                     CREATE TABLE $outputTableName AS SELECT a.*, (SELECT b.$idColumnUp 
                     FROM $inputUpperScaleTableName b WHERE a.$GEOMETRIC_COLUMN_LOW && b.$GEOMETRIC_COLUMN_UP AND 
                     ST_INTERSECTS(st_force2d(a.$GEOMETRIC_COLUMN_LOW), st_force2d(b.$GEOMETRIC_COLUMN_UP)) ORDER BY 
                     ST_AREA(ST_INTERSECTION(st_force2d(st_makevalid(a.$GEOMETRIC_COLUMN_LOW)), st_force2d(st_makevalid(b.$GEOMETRIC_COLUMN_UP)))) 
                     DESC LIMIT 1) AS $idColumnUp FROM $inputLowerScaleTableName a """
```

File: geoindicators/src/main/groovy/org/orbisgis/orbisprocess/geoclimate/geoindicators/SpatialUnits.groovy (hunk 1 above)
Comment (j3r3m1) [reply geoindicators/src/main/groovy/org/orbisgis/orbisprocess/geoclimate/geoindicators/SpatialUnits.groovy:390]: ok

File: geoindicators/src/main/groovy/org/orbisgis/orbisprocess/geoclimate/geoindicators/SpatialUnits.groovy (hunk 2 above)
Comment (j3r3m1) [reply geoindicators/src/main/groovy/org/orbisgis/orbisprocess/geoclimate/geoindicators/SpatialUnits.groovy:403]: ok

File: geoindicators/src/main/groovy/org/orbisgis/orbisprocess/geoclimate/geoindicators/SpatialUnits.groovy (hunk 3 above)
Comment (j3r3m1) [reply geoindicators/src/main/groovy/org/orbisgis/orbisprocess/geoclimate/geoindicators/SpatialUnits.groovy:421]: I have tried adding to the query:
```sql
(SELECT ST_AREA(ST_INTERSECTION(st_force2d(st_makevalid(a.$GEOMETRIC_COLUMN_LOW)), st_force2d(st_makevalid(b.$GEOMETRIC_COLUMN_UP)))) AS AREA 
                     FROM $inputUpperScaleTableName b WHERE a.$GEOMETRIC_COLUMN_LOW && b.$GEOMETRIC_COLUMN_UP AND 
                     ST_INTERSECTS(st_force2d(a.$GEOMETRIC_COLUMN_LOW), st_force2d(b.$GEOMETRIC_COLUMN_UP)) ORDER BY 
                     ST_AREA(ST_INTERSECTION(st_force2d(st_makevalid(a.$GEOMETRIC_COLUMN_LOW)), st_force2d(st_makevalid(b.$GEOMETRIC_COLUMN_UP)))) 
                     DESC LIMIT 1) AS $idColumnUp```

but I got a "multiple row issue"

File: geoindicators/src/main/groovy/org/orbisgis/orbisprocess/geoclimate/geoindicators/SpatialUnits.groovy [hunk 4]
```diff
@@ -351,53 +351,85 @@ IProcess createBlocks() {
...
+                                            ST_INTERSECTS(ST_POINTONSURFACE(a.$GEOMETRIC_COLUMN_LOW), st_force2d(b.$GEOMETRIC_COLUMN_UP))"""
+            }
+            else {
+                def temporaryTab = "temporaryTab"
+                if(nbRelations!=null){
+                    subqueryNbRelations += """DESC LIMIT $nbRelations"""
+                }
+                query += """DROP TABLE IF EXISTS $temporaryTab;
+                            CREATE TABLE $temporaryTab 
+                                    AS SELECT   a.*, b.$idColumnTab2,
```
Comment (ebocher) [inline geoindicators/src/main/groovy/org/orbisgis/orbisprocess/geoclimate/geoindicators/SpatialUnits.groovy:414]: Drop the temporary table

**Comment by ebocher** [comment]
This pr is for my "pomme"

**Pull request merged**